File: ports/ruby/lib/unicode_ruby/security/form/nfc_idempotence_witness.rb
# frozen_string_literal: true

require_relative "../identity/ucd"

module UnicodeRuby
  module Security
    module Form
      # NFC-idempotence-witness detection — inputs not already in NFC (or,
      # failing that, not in NFKC), the silent normalization-drift class.
      # Reports the first divergent position: a mismatch against NFC is
      # NonNfcForm; a sequence already in NFC but not NFKC is NonNfkcCompatForm.
      module NfcIdempotenceWitness
        Detection = Struct.new(:sub, :positions)

        module_function

        def first_divergence(a, b)
          common = [a.length, b.length].min
          (0...common).each do |i|
            return i if a[i] != b[i]
          end
          return common if a.length != b.length

          nil
        end

        def detect(input)
          nfc = Ucd.to_nfc(input)
          pos = first_divergence(input, nfc)
          return Detection.new("NonNfcForm", [pos]) unless pos.nil?

          nfkc = Ucd.to_nfkc(input)
          pos = first_divergence(input, nfkc)
          return Detection.new("NonNfkcCompatForm", [pos]) unless pos.nil?

          Detection.new(nil, [])
        end
      end
    end
  end
end

File: ports/lua/src/unicode_lua/security/form/nfc_idempotence_witness.lua
local ucd = require("unicode_lua.security.identity.ucd")

local M = {}

local function first_divergence(a, b)
  local common = math.min(#a, #b)
  for i = 1, common do
    if a[i] ~= b[i] then
      return i - 1
    end
  end
  if #a ~= #b then
    return common
  end
  return nil
end

function M.detect(input)
  local nfc = ucd.to_nfc(input)
  local pos = first_divergence(input, nfc)
  if pos ~= nil then
    return { sub = "NonNfcForm", positions = { pos } }
  end
  local nfkc = ucd.to_nfkc(input)
  pos = first_divergence(input, nfkc)
  if pos ~= nil then
    return { sub = "NonNfkcCompatForm", positions = { pos } }
  end
  return { sub = nil, positions = {} }
end

return M
